File: src/MagnusAddon/jaslib/language_services/janome_ex/tokenizing/godan_dictionary_form_stem.py
from __future__ import annotations

from typing import TYPE_CHECKING, override

from autoslot import Slots
from jaslib.language_services.janome_ex.tokenizing.analysis_token import IAnalysisToken

if TYPE_CHECKING:
    from jaslib.language_services.janome_ex.tokenizing.jn_token import JNToken

# this is the interface that is actually used in the text analysis pipeline

class SplitTokenBase(IAnalysisToken, Slots):
    def __init__(self,
                 source: JNToken,
                 surface: str,
                 base: str) -> None:
        self.source: JNToken = source
        self._surface: str = surface
        self._base_form: str = base

    @property
    @override
    def source_token(self) -> JNToken: return self.source
    @property
    @override
    def surface(self) -> str: return self._surface
    @property
    @override
    def base_form(self) -> str: return self._base_form

    @override
    def __repr__(self) -> str:
        return f"{self.__class__.__name__}('{self.surface}', '{self.base_form}')"

class DictionaryFormsTokenBase(SplitTokenBase, Slots):
    def __init__(self, source: JNToken, surface: str, base: str) -> None:
        super().__init__(source, surface, base)

    @property
    @override
    def is_inflectable_word(self) -> bool: return True  # todo: this feels odd, but without it it seems that things go haywire...

class DictionaryFormStem(DictionaryFormsTokenBase, Slots):
    def __init__(self, source: JNToken, surface: str, base: str) -> None:
        super().__init__(source, surface, base)

    @property
    @override
    def is_dictionary_verb_form_stem(self) -> bool: return True

class DictionaryFormInflection(DictionaryFormsTokenBase, Slots):
    def __init__(self, source: JNToken, surface: str, base: str) -> None:
        super().__init__(source, surface, base)

    @property
    @override
    def is_dictionary_verb_inflection(self) -> bool: return True

class GodanDictionaryFormStem(DictionaryFormStem, Slots):
    def __init__(self, source: JNToken, surface: str, base: str) -> None:
        super().__init__(source, surface, base)

class GodanPotentialDictionaryFormStem(GodanDictionaryFormStem, Slots):
    def __init__(self, source: JNToken, surface: str, base: str) -> None:
        super().__init__(source, surface, base)

    @property
    @override
    def is_godan_potential_stem(self) -> bool: return True

    @property
    @override
    def is_godan_verb(self) -> bool: return True  # todo this leaves us with, these tokens claiming both to be an ichidan(via the base class) and a godan... We can't tell which, but still..

class GodanDictionaryFormInflection(DictionaryFormInflection, Slots):
    def __init__(self, source: JNToken, surface: str, base: str) -> None:
        super().__init__(source, surface, base)

class GodanPotentialInflectionDictionaryFormStem(DictionaryFormStem, Slots):
    def __init__(self, source: JNToken, surface: str, base: str) -> None:
        super().__init__(source, surface, base)

    @property
    @override
    def is_godan_potential_inflection(self) -> bool: return True

class GodanPotentialInflectionDictionaryFormInflection(DictionaryFormInflection, Slots):
    def __init__(self, source: JNToken) -> None:
        super().__init__(source, "る", "る")

class IchidanDictionaryFormStem(DictionaryFormStem, Slots):
    def __init__(self, source: JNToken, surface: str, base: str) -> None:
        super().__init__(source, surface, base)

class IchidanDictionaryFormInflection(DictionaryFormInflection, Slots):
    def __init__(self, source: JNToken) -> None:
        super().__init__(source, "る", "る")

class KuruVerbDictionaryFormStem(DictionaryFormStem, Slots):
    def __init__(self, source: JNToken, surface: str, base: str) -> None:
        super().__init__(source, surface, base)

class KuruVerbDictionaryFormInflection(DictionaryFormInflection, Slots):
    def __init__(self, source: JNToken) -> None:
        super().__init__(source, "る", "る")

class SuruVerbDictionaryFormStem(DictionaryFormStem, Slots):
    def __init__(self, source: JNToken, surface: str, base: str) -> None:
        super().__init__(source, surface, base)

class SuruVerbDictionaryFormInflection(DictionaryFormInflection, Slots):
    def __init__(self, source: JNToken) -> None:
        super().__init__(source, "る", "る")

File: src/src_dotnet/JAStudio.Core/LanguageServices/JanomeEx/Tokenizing/GodanDictionaryFormStem.cs
namespace JAStudio.Core.LanguageServices.JanomeEx.Tokenizing;

// Base class for split tokens
public abstract class SplitTokenBase : IAnalysisToken
{
   JNToken Source { get; }
   readonly string _surface;
   readonly string _baseForm;

   protected SplitTokenBase(JNToken source, string surface, string baseForm)
   {
      Source = source;
      _surface = surface;
      _baseForm = baseForm;
   }

   public override JNToken SourceToken => Source;
   public override string Surface => _surface;
   public override string BaseForm => _baseForm;

   public override string ToString() => $"{GetType().Name}('{Surface}', '{BaseForm}')";
}

// Base class for dictionary form tokens
public abstract class DictionaryFormsTokenBase : SplitTokenBase
{
   protected DictionaryFormsTokenBase(JNToken source, string surface, string baseForm)
      : base(source, surface, baseForm) {}

   // TODO: this feels odd, but without it it seems that things go haywire...
   public override bool IsInflectableWord => true;
}

// Dictionary form stem
public class DictionaryFormStem : DictionaryFormsTokenBase
{
   public DictionaryFormStem(JNToken source, string surface, string baseForm)
      : base(source, surface, baseForm) {}

   public override bool IsDictionaryVerbFormStem => true;
}

// Dictionary form inflection
public class DictionaryFormInflection : DictionaryFormsTokenBase
{
   public DictionaryFormInflection(JNToken source, string surface, string baseForm)
      : base(source, surface, baseForm) {}

   public override bool IsDictionaryVerbInflection => true;
}

// Godan dictionary form stem
public class GodanDictionaryFormStem : DictionaryFormStem
{
   public GodanDictionaryFormStem(JNToken source, string surface, string baseForm)
      : base(source, surface, baseForm) {}
}

// Godan potential dictionary form stem
public class GodanPotentialDictionaryFormStem : GodanDictionaryFormStem
{
   public GodanPotentialDictionaryFormStem(JNToken source, string surface, string baseForm)
      : base(source, surface, baseForm) {}

   public override bool IsGodanPotentialStem => true;

   // TODO: this leaves us with these tokens claiming both to be an ichidan (via the base class) and a godan... We can't tell which, but still..
   public override bool IsGodanVerb => true;
}

// Godan dictionary form inflection
public class GodanDictionaryFormInflection : DictionaryFormInflection
{
   public GodanDictionaryFormInflection(JNToken source, string surface, string baseForm)
      : base(source, surface, baseForm) {}
}

// Godan potential inflection dictionary form stem
public class GodanPotentialInflectionDictionaryFormStem : DictionaryFormStem
{
   public GodanPotentialInflectionDictionaryFormStem(JNToken source, string surface, string baseForm)
      : base(source, surface, baseForm) {}

   public override bool IsGodanPotentialInflection => true;
}

// Godan potential inflection dictionary form inflection
public class GodanPotentialInflectionDictionaryFormInflection : DictionaryFormInflection
{
   public GodanPotentialInflectionDictionaryFormInflection(JNToken source)
      : base(source, "る", "る") {}
}

// Ichidan dictionary form stem
public class IchidanDictionaryFormStem : DictionaryFormStem
{
   public IchidanDictionaryFormStem(JNToken source, string surface, string baseForm)
      : base(source, surface, baseForm) {}
}

// Ichidan dictionary form inflection
public class IchidanDictionaryFormInflection : DictionaryFormInflection
{
   public IchidanDictionaryFormInflection(JNToken source)
      : base(source, "る", "る") {}
}

// Kuru verb dictionary form stem
public class KuruVerbDictionaryFormStem : DictionaryFormStem
{
   public KuruVerbDictionaryFormStem(JNToken source, string surface, string baseForm)
      : base(source, surface, baseForm) {}
}

// Kuru verb dictionary form inflection
public class KuruVerbDictionaryFormInflection : DictionaryFormInflection
{
   public KuruVerbDictionaryFormInflection(JNToken source)
      : base(source, "る", "る") {}
}

// Suru verb dictionary form stem
public class SuruVerbDictionaryFormStem : DictionaryFormStem
{
   public SuruVerbDictionaryFormStem(JNToken source, string surface, string baseForm)
      : base(source, surface, baseForm) {}
}

// Suru verb dictionary form inflection
public class SuruVerbDictionaryFormInflection : DictionaryFormInflection
{
   public SuruVerbDictionaryFormInflection(JNToken source)
      : base(source, "る", "る") {}
}
